src/tests/payments.js

```diff
@@ -0,0 +1,277 @@
+    }
+    return this.promiseToPost;
+};
+
+FakeRequests.prototype.get = function() {
+    // When the Fireplace adapter makes a get request it is checking the
+    // status of the transaction.
+    var response = {
+        // The transaction has been verified and has completed successfully.
+        status: 'complete',
+    };
+    return defer.Deferred().resolve(response).promise();
+};
+
+
+function fakeDeps() {
+    return {
+        requests: new FakeRequests(),
+        notification: {notification: function() {}},
+        settings: {},
+    };
+}
+
+
+function PaymentWindow() {
+    this.location = '';
+    this.closed = false;
+    this.close = function() {};
+}
+
+
+function FakeFireplaceWindow() {
+    this.origin = 'https://marketplace.firefox.com';
+    this.open = function() {
+        this.paymentWindow = new PaymentWindow();
+        return this.paymentWindow;
+    };
+    this.addEventListener = function(eventType, handler) {
+        logger.log('calling handler for:', eventType);
+        // Simulate Webpay's postMessage back to fxpay.
+        handler({
+            origin: this.origin,
+            data: {
+                // Indicate that the user payment flow is
+                // finished. Next fxpay will verify the transaction.
+                status: 'ok',
+            }
+        });
+    };
+    this.removeEventListener = function() {};
+}
+
+
+function FakeProduct() {
+    this.slug = 'some-slug';
+    this.payment_required = true;
+}
+
+
+function purchaseOptions(opt) {
+    opt = opt || {};
+    return {
+        fxpaySettings: {
+            mozApps: null,
+            mozPay: null,
+            window: opt.fakeWindow || new FakeFireplaceWindow(),
+            // On Firefox there is some difference in fxpay's scope
+            // that causes fxpay.init() not to clear this. Hmm.
+            initError: null,
+        }
+    };
+}
+
+
+test('payments: complete payment successfully', function(done, fail) {
+    mock(
+        'payments',
+        fakeDeps(),
+        function(payments) {
+            var product = new FakeProduct();
+            payments.purchase(product, purchaseOptions()).done(function() {
+                logger.log('payment.purchase() is done');
+                done();
+            }).fail(function(_, product, reason) {
+                logger.error('payment.purchase() failed:', reason);
+                fail(reason);
+            });
+        },
+        fail
+    );
+});
+
+test('payments: product already purchased', function(done, fail) {
+    var deps = fakeDeps();
+
+    // Reject the JWT post with a 409 response.
+    deps.requests.promiseToPost = defer.Deferred().reject({}, null, 409).promise();
+
+    mock(
+        'payments',
+        deps,
+        function(payments) {
+            var product = new FakeProduct();
+            payments.purchase(product, purchaseOptions()).done(function() {
+                logger.log('payment.purchase() is done');
```

Comment:
the thing is, it does not get the error. That error is intercepted and the product is resolved as if it was successful by any other means, i.e. it looks no different from a successful purchase caused by a real transaction. I got tripped up on this too but let me see if there is a way to distinguish it.